src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -677,6 +742,234 @@ impl Crumbable for crate::Block<Ast> {
+                seg.body    = set_pattern(&seg.body,val,new_ast)?
+            },
+            MatchCrumb::Segs{index,val:SegmentMatchCrumb::Head} => {
+                let mut seg = new_self.segs.get_mut_or_err(*index,"segment")?;
+                seg.head    = new_ast
+            }
+        }
+        Ok(new_self)
+
+    }
+
+    fn iter_subcrumbs<'a>(&'a self) -> Box<dyn Iterator<Item=Self::Crumb> + 'a> {
+
+        fn iter_pat(pat:&MacroPatternMatch<Shifted<Ast>>) -> Vec<Vec<PatternMatchCrumb>> {
+            use crate::MacroPatternMatchRaw::*;
+
+            let mut crumbs   = vec![];
+            let mut patterns = vec![(vec![],pat)];
+            while let Some((mut crumb,pattern)) = patterns.pop() {
+                match pattern.deref() {
+                    Begin(_)   => (),
+                    End(_)     => (),
+                    Nothing(_) => (),
+                    Build(_)   => {crumb.push(PatternMatchCrumb::Build)  ; crumbs.push(crumb)},
+                    Err(_)     => {crumb.push(PatternMatchCrumb::Err)    ; crumbs.push(crumb)},
+                    Tok(_)     => {crumb.push(PatternMatchCrumb::Tok)    ; crumbs.push(crumb)},
+                    Blank(_)   => {crumb.push(PatternMatchCrumb::Blank)  ; crumbs.push(crumb)},
+                    Var(_)     => {crumb.push(PatternMatchCrumb::Var)    ; crumbs.push(crumb)},
+                    Cons(_)    => {crumb.push(PatternMatchCrumb::Cons)   ; crumbs.push(crumb)},
+                    Opr(_)     => {crumb.push(PatternMatchCrumb::Opr)    ; crumbs.push(crumb)},
+                    Mod(_)     => {crumb.push(PatternMatchCrumb::Mod)    ; crumbs.push(crumb)},
+                    Num(_)     => {crumb.push(PatternMatchCrumb::Num)    ; crumbs.push(crumb)},
+                    Text(_)    => {crumb.push(PatternMatchCrumb::Text)   ; crumbs.push(crumb)},
+                    Block(_)   => {crumb.push(PatternMatchCrumb::Block)  ; crumbs.push(crumb)},
+                    Macro(_)   => {crumb.push(PatternMatchCrumb::Macro)  ; crumbs.push(crumb)},
+                    Invalid(_) => {crumb.push(PatternMatchCrumb::Invalid); crumbs.push(crumb)},
+                    Except(pat) => {
+                        crumb.push(PatternMatchCrumb::Except);
+                        patterns.push((crumb,&pat.elem))
+                    },
+                    Tag(pat) => {
+                        crumb.push(PatternMatchCrumb::Tag);
+                        patterns.push((crumb,&pat.elem))
+                    },
+                    Cls(pat) => {
+                        crumb.push(PatternMatchCrumb::Cls);
+                        patterns.push((crumb,&pat.elem))
+                    },
+                    Or(pat) => {
+                        crumb.push(PatternMatchCrumb::Or);
+                        match &pat.elem {
+                            crate::Switch::Left (pat) => patterns.push((crumb,&pat.value)),
+                            crate::Switch::Right(pat) => patterns.push((crumb,&pat.value)),
```

Comment:
I think that `Switch` should provide accessors to its contents.
Perhaps implement `Deref` for it. (esentially Switch<T> is T + flag, so it could provide access to the T)